penrose-web/src/compiler/Style.ts

```diff
@@ -1,7 +1,2068 @@
+  const selEnvs: SelEnv[] = prog.map(e => {
+    const res = checkHeader(varEnv, e.header);
+    // Put selector AST in just for debugging
+    res.header = { tag: "Just", contents: e.header };
+    return res;
+  });
+  // const errors = ... TODO(errors)
+  return selEnvs; // TODO
+};
+
+//#endregion
+
+//#region Types and code for finding substitutions
+
+// Judgment 20. A substitution for a selector is only correct if it gives exactly one
+//   mapping for each Style variable in the selector. (Has test)
+export const fullSubst = (selEnv: SelEnv, subst: Subst): boolean => {
+  // Check if a variable is a style variable, not a substance one
+  const isStyVar = (e: string): boolean => selEnv.varProgTypeMap[e][0].tag === "StyProgT";
+
+  // Equal up to permutation (M.keys ensures that there are no dups)
+  const selStyVars = Object.keys(selEnv.sTypeVarMap).filter(isStyVar);
+  const substStyVars = Object.keys(subst);
+  // Equal up to permutation (keys of an object in js ensures that there are no dups)
+  return _.isEqual(selStyVars.sort(), substStyVars.sort());
+};
+
+// Check that there are no duplicate keys or vals in the substitution (Has test)
+export const uniqueKeysAndVals = (subst: Subst): boolean => {
+  // All keys already need to be unique in js, so only checking values
+  const vals = Object.values(subst);
+  const valsSet = {};
+
+  for (let i = 0; i < vals.length; i++) {
+    valsSet[vals[i]] = 0; // This 0 means nothing, we just want to create a set of values
+  }
+
+  // All entries were unique if length didn't change (ie the nub didn't change)
+  return Object.keys(valsSet).length === vals.length;
+};
+
+const couldMatchRels = (typeEnv: Env, rels: RelationPattern[], stmt: SubStmt): boolean => {
+  // TODO < (this is an optimization; will only implement if needed)
+  return true;
+};
+
+//#region (subregion? TODO fix) Applying a substitution
+//// Apply a substitution to various parts of Style (relational statements, exprs, blocks)
+
+// Recursively walk the tree, looking up and replacing each Style variable encountered with a Substance variable
+// If a Sty var doesn't have a substitution (i.e. substitution map is bad), keep the Sty var and move on
+// COMBAK: return "maybe" if a substitution fails?
+// COMBAK: Add a type for `lv`? It's not used here
+const substituteBform = (lv: any, subst: Subst, bform: BindingForm): BindingForm => {
+  // theta(B) = ...
+  if (bform.tag === "SubVar") {
+    // Variable in backticks in block or selector (e.g. `X`), so nothing to substitute
+    return bform;
+  } else if (bform.tag === "StyVar") {
+    // Look up the substitution for the Style variable and return a Substance variable
+    // Returns result of mapping if it exists (y -> x)
+    const res = subst[bform.contents.value];
+
+    if (res) {
+      return {
+        ...bform, // Copy the start/end loc of the original Style variable, since we don't have Substance parse info (COMBAK)
```

Comment:
Again, the `ASTNode` type is not really designed for this. You could do a postprocessing step to throw away all node-specific data.